src/server/kernel/wseat.cpp

```diff
@@ -960,6 +965,138 @@ void WSeat::notifyFrame(WCursor *cursor)
     d->doNotifyFrame();
 }
 
+void WSeat::notifySwipeBegin(WCursor *cursor, WInputDevice *device, uint32_t time_msec, uint32_t fingers)
+{
+    W_D(WSeat);
+    if (d->gestureState) {
+        qCWarning(qLcWlrGestureEvents) << "Unexpected SwipeBegin while already active";
+    }
+    d->gestureState = true;
+    d->gestureFingers = fingers;
+    auto qwDevice = qobject_cast<QPointingDevice*>(device->qtDevice());
+    auto *w = cursor->eventWindow();
+    const QPointF &global = cursor->position();
+    const QPointF local = w ? global - QPointF(w->position()) : QPointF();
+    QWindowSystemInterface::handleGestureEvent(w, time_msec, qwDevice,
+        Qt::NativeGestureType::BeginNativeGesture, local, global, fingers);
+
+    Q_EMIT sendSwipeBegin(time_msec, fingers);
+}
+
+void WSeat::notifySwipeUpdate(WCursor *cursor, WInputDevice *device, uint32_t time_msec, const QPointF &pos)
+{
+    W_D(WSeat);
+    if (!d->gestureState) {
+        qCWarning(qLcWlrGestureEvents) << "Unexpected SwipeUpdate while not begin";
+        return;
+    }
+    auto qwDevice = qobject_cast<QPointingDevice*>(device->qtDevice());
+    auto *w = cursor->eventWindow();
+    const QPointF &global = cursor->position();
+    const QPointF local = w ? global - QPointF(w->position()) : QPointF();
+    QWindowSystemInterface::handleGestureEventWithValueAndDelta(w, time_msec, qwDevice,
```

Comment:
Why is not `SwipeNativeGesture`?

Reply:
see 
https://github.com/qt/qtwayland/blob/a5bea434aa9137f9b0ba6170fbb5a237e890b021/src/client/qwaylandwindow.cpp#L1268

Reply:
Maybe is a Qt bug.